src/audio_pixel.cpp

```diff
@@ -0,0 +1,183 @@
+#include "audio_pixel.h"
+
+// ************* audio_pixel_block_t ****************
+
+void audio_pixel_block_t::update(std::vector<double>& sample_buffer, int num_channels, int sample_rate) {
+    // constructs an audio pixel block using interleaved sample buffer and num channels
+    // clear the current set of audio pixels 
+    m_channel_pixels.clear();
+
+    // initalize the m_channel_pixels to have empty audio pixel vectors
+    for (int i = 0; i < num_channels; i++) {
+        std::vector<audio_pixel_t> empty_pixel_vec;
+        m_channel_pixels.push_back(empty_pixel_vec);
+    }
+
+    //calcualte the samples per audio pixel and initalize an audio pixel
+    int samples_per_pixel = m_pix_per_s * sample_rate;
+    audio_pixel_t curr_pixel = audio_pixel_t(0, -1, 0);
+
+    for (int channel= 0; channel < num_channels; channel++) {
+        int collected_samples = 0;
+
+        for (int i = 0; i < sample_buffer.size(); i += num_channels) { 
+            double curr_sample = sample_buffer[i];
+
+            // increment collected samples after creating new audio pixel 
+            collected_samples ++;
+    
+            // update all fields of the current pixeld based on the current sample
+            curr_pixel.m_max = std::max(curr_pixel.m_max, curr_sample);
+            curr_pixel.m_min = std::min(curr_pixel.m_min, curr_sample);
+            curr_pixel.m_rms += (curr_sample * curr_sample);
+
+            if ((collected_samples % samples_per_pixel) == 0) {
+                // complete the RMS calculation, this method accounts of edge cases (total sample % samples per pixel != 0)
+                curr_pixel.m_rms = curr_pixel.m_rms / collected_samples;
+
+                // add the curr pixel to its channel in m_channel_pixels, clear curr_pixel
+                m_channel_pixels[channel].push_back(curr_pixel);
+
+                // set max to equal reaper min, and min to equal reaper max
+                curr_pixel = audio_pixel_t(DBL_MIN, DBL_MAX, 0);
+            }
+        }
+    }
+}
+
+// ************* audio_pixel_mipmap_t ****************
+
+audio_pixel_block_t audio_pixel_mipmap_t::get_pixels(double t0, double t1, double pix_per_s) {
+    // start by grabbing pixels from the nearest two pixel blocks TODO: this will be a vector<double>
+    double nearest_pps = get_nearest_pps(pix_per_s);
+
+    if (nearest_pps == pix_per_s) // last edge case, when hte given pps is already a block
+        return m_blocks[nearest_pps];
```

Comment:
do we wanna return the entire block here? or just a block for the specified time range?

Reply:
i gotcha, i'll fix this to call the method for a region

Reply:
const std::vector<std::vector<audio_pixel_t>>& this will be updated to `const 2daudiopixeltype`